ee/api/authentication.py

```diff
@@ -242,3 +252,157 @@ def get_user_id(self, details, response):
+    """
+    Implements Vercel Marketplace API authentication.
+    This authentication uses the OpenID Connect Protocol (OIDC).
+    Vercel sends a JSON web token (JWT) signed with Vercel’s private key and verifiable
+    using Vercel’s public JSON Web Key Sets (JWKS) available through VERCEL_JWKS_URL
+
+    For detailed reference of User/System Auth OIDC token claims schema, see:
+    https://vercel.com/docs/integrations/create-integration/marketplace-api#marketplace-partner-api-authentication
+    """
+
+    VercelAuthType = Literal["user", "system"]
+
+    VERCEL_AUTH_TYPES: tuple[VercelAuthType, ...] = ("user", "system")
+    USER_SUB_RE = re.compile(r"^account:[0-9a-fA-F]+:user:[0-9a-fA-F]+$")
+    SYSTEM_SUB_RE = re.compile(r"^account:[0-9a-fA-F]+$")
+    VERCEL_ISSUER = "https://marketplace.vercel.com"
+
+    def authenticate(self, request: Request) -> tuple[VercelUser, None] | None:
+        token = self._get_bearer_token(request)
+        if not token:
+            return None
+
+        auth_type = self._get_vercel_auth_type(request)
+
+        try:
+            payload = self._validate_jwt_token(token, auth_type)
+            logger.info("Vercel auth successful", auth_type=auth_type, account_id=payload.get("account_id"))
+            return VercelUser(claims=payload), None
+        except jwt.InvalidTokenError as e:
+            logger.warning("Vercel auth failed", auth_type=auth_type, error=str(e))
+            raise AuthenticationFailed(f"Invalid {auth_type} authentication token")
+        except Exception as e:
+            logger.exception("Vercel auth error", auth_type=auth_type, error=str(e))
+            raise AuthenticationFailed(f"{auth_type.title()} authentication failed")
+
+    def _get_bearer_token(self, request: Request) -> str | None:
+        if auth_header := request.META.get("HTTP_AUTHORIZATION"):
+            parts = auth_header.split()
+            if len(parts) == 2 and parts[0].lower() == "bearer":
+                return parts[1]
+
+        return None
+
+    def _get_vercel_auth_type(self, request: Request) -> "VercelAuthentication.VercelAuthType":
+        auth_type = request.headers.get("X-Vercel-Auth", "").lower()
+
+        if auth_type not in self.VERCEL_AUTH_TYPES:
+            raise AuthenticationFailed("Missing or invalid X-Vercel-Auth header")
+
+        return cast("VercelAuthentication.VercelAuthType", auth_type)
+
+    def _validate_jwt_token(self, token: str, auth_type: "VercelAuthentication.VercelAuthType") -> VercelClaims:
+        payload = self._decode_token(token)
+        return self._validate_claims(payload, auth_type)
+
+    def _validate_claims(
+        self, payload: dict[str, Any], auth_type: "VercelAuthentication.VercelAuthType"
+    ) -> VercelClaims:
+        required_claims = ["iss", "sub", "aud"]
+
+        for claim in required_claims:
+            if claim not in payload:
+                raise jwt.InvalidTokenError(f"Missing required claim: {claim}")
```

Comment:
nit: do we want to add `account_id` and `installation_id` to this? They're required in both System Authentication and User Authentication.